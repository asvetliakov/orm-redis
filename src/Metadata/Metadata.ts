/**
 * Metadata key to mark class to be stored as redis hash
 */
export const REDIS_ENTITY = Symbol("Redis Hash");

/**
 * Metadata key to store redis properties metadatas
 */
export const REDIS_PROPERTIES = Symbol("Redis properties");

/**
 * Metadata key to keep serialized initial value of property from redis. For collections this will be link to the corresponding hashmap/set
 */
export const REDIS_VALUE = Symbol("Redis initial value");

/**
 * Metadata key to keep the initial state of redis collections (map/set)
 */
export const REDIS_COLLECTION_VALUE = Symbol("Redis initial collection value");


export type PropertyMetadata = SimplePropertyMetadata | RelationPropertyMetadata;

/**
 * Hash simple property metadata
 * 
 * @export
 * @interface SimplePropertyMetadata
 */
export interface SimplePropertyMetadata {
    /**
     * Property name in class
     */
    propertyName: string;
    /**
     * Property name stored in redis
     */
    propertyRedisName: string;
    /**
     * Property type
     */
    propertyType: any;
    /**
     * True if property must be identify column
     */
    isIdentifyColumn: boolean;
    /**
     * Relation flag
     */
    isRelation: false;
}

/**
 * Relation options
 * 
 * @export
 * @interface RelationOptions
 */
export interface RelationOptions {
    /**
     * Property name in redis
     */
    propertyName?: string;
    /**
     * True to automatically create new hash for relation
     */
    cascadeInsert?: boolean;
    /**
     * True to automatically update corresponding relation hash
     */
    cascadeUpdate?: boolean;
}

/**
 * Hash relation property metadata
 * 
 * @export
 * @interface RelationPropertyMetadata
 */
export interface RelationPropertyMetadata {
    /**
     * Property name in class
     */
    propertyName: string;
    /**
     * Property name stored in redis
     */
    propertyRedisName: string;
    /**
     * Property type.
     * For set property relation it will be Set itself
     * For single property relation it will be equal to relationType
     */
    propertyType: any;
    /**
     * Relation flag
     */
    isRelation: true;
    /**
     * Relation type. Must be type of related entity
     */
    relationType: any;
    /**
     * Relation options
     */
    relationOptions: RelationOptions;
}

/**
 * Return true if given object is redis entity like
 * 
 * @export
 * @param entity 
 * @returns 
 */
export function isRedisEntity(entity: object): boolean {
    if (typeof entity !== "object" || entity === null) {
        return false;
    }
    const metadata = Reflect.getMetadata(REDIS_ENTITY, entity.constructor);
    return !!metadata;
}

/**
 * Return entity name
 * 
 * @export
 * @param entityOrEntityClass 
 * @returns 
 */
export function getEntityName(entityOrEntityClass: object | Function): string | undefined {
    const metadata = Reflect.getMetadata(REDIS_ENTITY, typeof entityOrEntityClass === "function" ? entityOrEntityClass : entityOrEntityClass.constructor);
    return metadata;
}

/**
 * Return entity id
 * 
 * @export
 * @param entity 
 * @returns 
 */
export function getEntityId(entity: { [key: string]: any }): string | number | undefined {
    if (!isRedisEntity(entity)) {
        return undefined;
    }
    const metadatas: PropertyMetadata[]  = Reflect.getMetadata(REDIS_PROPERTIES, entity.constructor);
    const idMetadata = metadatas && metadatas.find(val => !val.isRelation && val.isIdentifyColumn);
    if (!idMetadata) {
        return undefined;
    }
    if (typeof entity[idMetadata.propertyName] === "number" || typeof entity[idMetadata.propertyName] === "string") {
        return entity[idMetadata.propertyName];
    }
    return undefined;
}

/**
 * Returns redis properties metadata
 * 
 * @export
 * @param entityOrEntityClass 
 * @returns 
 */
export function getEntityProperties(entityOrEntityClass: object | Function): PropertyMetadata[] | undefined {
    return Reflect.getMetadata(REDIS_PROPERTIES, typeof entityOrEntityClass === "function" ? entityOrEntityClass : entityOrEntityClass.constructor);
}

/**
 * Return full hash id
 * 
 * @export
 * @param entityOrEntityClass 
 * @param [hashId]
 * @returns 
 */
export function getEntityFullId(entityOrEntityClass: object | Function, hashId?: string | number): string | undefined {
    const name = getEntityName(entityOrEntityClass);
    const id = typeof hashId !== "undefined"
        ? hashId
        : typeof entityOrEntityClass === "object"
            ? getEntityId(entityOrEntityClass)
            : undefined;        
    if (name && typeof id !== "undefined") {
        return `e:${name}:${id}`;
    }
}