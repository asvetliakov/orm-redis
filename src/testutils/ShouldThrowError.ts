
export class ShouldThrowError extends Error {
    public constructor() {
        super("The test should throw");
    }
}