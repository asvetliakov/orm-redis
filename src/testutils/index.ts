export * from "./redis";
export * from "./RedisTestMonitor";