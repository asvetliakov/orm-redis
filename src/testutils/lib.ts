// See https://github.com/Microsoft/TypeScript/issues/16872
// USE ONLY IN TESTS

const MyMap = Map;
const MySet = Set;
const MyInt8Array = Int8Array;
const MyDate = Date;

export { MyMap as Map };
export { MySet as Set };
export { MyInt8Array as Int8Array };
export { MyDate as Date };