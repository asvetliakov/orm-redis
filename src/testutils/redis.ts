import { Connection, ConnectionOptions } from "../Connection/Connection";
import * as redis from "../utils/PromisedRedis";


/**
 * Get available database from pool of databases (presented as SET in redis)
 * 
 * @returns 
 */
export async function getAvailableDatabase(): Promise<string> {
    const client = createRawRedisClient();
    let dbNumber: string | null = await client.spopAsync("available_db");
    // Set doesn't exist
    if (!dbNumber) {
        // Create set of db connetions
        await client.saddAsync("available_db", ["1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15"]);
        // Obtain again
        dbNumber = await client.spopAsync("available_db");
    }
    if (!dbNumber) {
        throw new Error("Unable to obtain database number");
    }
    await client.quitAsync();
    return dbNumber;
}

/**
 * Release available database and return it to the pool
 * 
 * @export
 * @param dbNumber 
 * @returns 
 */
export async function releaseDatabase(dbNumber: string): Promise<void> {
    const client = redis.createClient(parseInt(process.env.REDIS_PORT!) || 6379, process.env.REDIS_HOST);
    await client.saddAsync("available_db", dbNumber);
    await client.quitAsync();
}

/**
 * Create redis client
 * 
 * @export
 * @returns 
 */
export function createRawRedisClient(): redis.RedisClient {
    const client = redis.createClient(parseInt(process.env.REDIS_PORT!) || 6379, process.env.REDIS_HOST);
    return client;
}


// Just in case if the global state will be shared between tests
const connectionToDbNumber = new WeakMap<Connection, string>();

/**
 * Create redis connection
 * 
 * @param [options={}]
 * @export
 * @returns 
 */
export async function createRedisConnection(options: ConnectionOptions = {}): Promise<Connection> {
    const dbNumber = await getAvailableDatabase();
    const connection = new Connection();
    connectionToDbNumber.set(connection, dbNumber);
    try {
        await connection.connect({
            db: dbNumber.toString(),
            host: process.env.REDIS_HOST,
            port: parseInt(process.env.REDIS_PORT!) || 6379,
            ...options
        });
    } catch (e) {
        await releaseDatabase(dbNumber);
        throw e;
    }
    return connection;
}

/**
 * Clean & release redis connection
 * 
 * @export
 * @param connection 
 * @returns 
 */
export async function cleanRedisConnection(connection: Connection): Promise<void> {
    if (connection.isConnected) {
        await connection.flushdb();
    }
    const dbNumber = connectionToDbNumber.get(connection);
    if (dbNumber) {
        await releaseDatabase(dbNumber);
    }
    if (connection.isConnected) {
        await connection.disconnect();
    }
}