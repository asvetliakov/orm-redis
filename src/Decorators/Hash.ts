import { REDIS_HASH } from "../Metadata/Metadata";

/**
 * Defines redis hash
 * 
 * @export
 * @param {string} [name] 
 * @returns void
 */
export function Hash(name?: string): ClassDecorator {
    return function (constructor: Function) {
        Reflect.defineMetadata(REDIS_HASH, name ? name : constructor.name, constructor);
    };
}