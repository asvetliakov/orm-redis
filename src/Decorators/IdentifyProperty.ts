import { PropertyMetadata, REDIS_PROPERTIES } from "../Metadata/Metadata";

/**
 * Defines identify property for given redis hash.
 * Hash must contain one IdentifyProperty
 * 
 * @export
 * @param type Type
 * @param [name] Redis property name
 * @returns 
 */
export function IdentifyProperty(nameOrType?: string | typeof Number | typeof String, type?: typeof Number | typeof String): PropertyDecorator {
    return function (target: Object, propertyKey: string): void {
        const designType: any = Reflect.getMetadata("design:type", target, propertyKey);
        const name = typeof nameOrType === "string" ? nameOrType : propertyKey;
        const propertyType = typeof nameOrType === "function"
            ? nameOrType
            : type
                ? type
                : designType;        

        if (propertyType !== String && propertyType !== Number) {
            throw new Error("Identify property must be string or number type");
        }
        const properties: PropertyMetadata[] = Reflect.getMetadata(REDIS_PROPERTIES, target.constructor) || [];
        properties.push({
            propertyName: propertyKey,
            propertyRedisName: name,
            isIdentifyColumn: true,
            propertyType: propertyType,
            isRelation: false
        });
        Reflect.defineMetadata(REDIS_PROPERTIES, properties, target.constructor);
    };
}