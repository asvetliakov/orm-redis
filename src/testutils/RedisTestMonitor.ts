import { RedisClient } from "redis";
import { Connection } from "../Connection/Connection";
import { createRawRedisClient, getDatabaseNumberForConnection } from "./redis";

export class RedisTestMonitor {
    // Array of calls -> results without any timestamps
    private _calls: string[][][] = [];

    /**
     * Redis client
     * 
     * @private
     */
    private client: RedisClient;

    /**
     * Filter calls for specific redis db (each test connection connects to different db)
     * 
     * @private
     */
    private filterDbNumber?: string;

    /**
     * All monitor calls in format [request, response]
     * 
     * @readonly
     */
    public get calls(): string[][][] {
        return this._calls;
    }

    /**
     * Request only calls
     * 
     * @readonly
     */
    public get requests(): string[][] {
        return this._calls.map(([request, result]) => request);
    }

    /**
     * Responses only
     * 
     * @readonly
     */
    public get responses(): string[][] {
        return this._calls.map(([request, result]) => result);
    }


    /**
     * Release monitor
     * 
     * @returns 
     */
    public async release(): Promise<void> {
        await this.client.quitAsync();
    }

    /**
     * Reset monitor calls
     * @param timeToWaitMs Wait some time before clearing
     * 
     */
    public async clearMonitorCalls(timeToWaitMs?: number): Promise<void> {
        if (timeToWaitMs) {
            await this.wait(timeToWaitMs);
        }
        this._calls = [];
    }

    /**
     * Wait some time
     * 
     * @param timeToWaitMs 
     * @returns 
     */
    public async wait(timeToWaitMs: number): Promise<void> {
        await new Promise<void>(resolve => setTimeout(resolve, timeToWaitMs));
    }

    /**
     * Logger
     * 
     * @protected
     * @param time 
     * @param commandArgs 
     * @param reply 
     * @returns 
     */
    protected monitor(time: number, commandArgs: string[], reply: string): void {
        const repliesArgs = reply.split(" "); // 1500637727.256744 [12 172.20.0.1:48986] \"multi\"
        const justReply = repliesArgs.splice(3).map(r => r.slice(1, -1));
        const dbNumber = repliesArgs[1].slice(1); // [12 172.20.0.1:48986]
        if (this.filterDbNumber) {
            if (!dbNumber) {
                return;
            }
            if (dbNumber !== this.filterDbNumber) {
                return;
            }
        }
        this._calls.push([commandArgs, justReply as any]);
    }

    /**
     * Create monitor
     * 
     * @static
     * @param [connection] Monitor only for requests/replies for given connection db
     * @returns 
     */
    public static async create(connection?: Connection): Promise<RedisTestMonitor> {
        const instance = new RedisTestMonitor();
        if (connection) {
            instance.filterDbNumber = getDatabaseNumberForConnection(connection);
        }
        instance.client = await createRawRedisClient();
        instance.client.on("monitor", instance.monitor.bind(instance));
        await instance.client.monitorAsync();
        return instance;
    }
}