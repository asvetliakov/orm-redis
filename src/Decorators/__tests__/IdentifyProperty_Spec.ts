import { REDIS_PROPERTIES } from "../../Metadata/Metadata";
import { ShouldThrowError } from "../../testutils/ShouldThrowError";
import { Hash } from "../Hash";
import { IdentifyProperty } from "../IdentifyProperty";

@Hash()
class Test {
    @IdentifyProperty(String)
    public id: string;
}

@Hash()
class Test2 {
    @IdentifyProperty("somename")
    public id: number;
}

class Invalid1 {
    public id: Date;
}

class Invalid2 {
    public id: Object;
}

it("Throws error if type of property is not number or string", () => {
    try {
        const test = new Invalid1();
        IdentifyProperty(Date as any)(test, "id");
        throw new ShouldThrowError();
    } catch (e) {
        if (e instanceof ShouldThrowError) { throw e; }
    }
    try {
        const test = new Invalid2();
        IdentifyProperty(Object as any)(test, "id");
        throw new ShouldThrowError();
    } catch (e) {
        if (e instanceof ShouldThrowError) { throw e; }
    }
});

it("Defines metadata without custom name", () => {
    const metadata = Reflect.getMetadata(REDIS_PROPERTIES, Test);
    expect(metadata).toMatchSnapshot();
});

it("Defines metadata with custom name", () => {
    const metadata = Reflect.getMetadata(REDIS_PROPERTIES, Test2);
    expect(metadata).toMatchSnapshot();
});