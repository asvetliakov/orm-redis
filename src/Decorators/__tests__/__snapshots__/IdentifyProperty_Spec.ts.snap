// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Defines metadata with custom name 1`] = `
Array [
  Object {
    "isIdentifyColumn": true,
    "isRelation": false,
    "propertyName": "id",
    "propertyRedisName": "somename",
    "propertyType": [Function],
  },
]
`;

exports[`Defines metadata without custom name 1`] = `
Array [
  Object {
    "isIdentifyColumn": true,
    "isRelation": false,
    "propertyName": "id",
    "propertyRedisName": "id",
    "propertyType": [Function],
  },
]
`;
